x-pack/plugin/apm: introduce x-pack-apm plugin

axw opened the pull request:
Introduce the `x-pack-apm` plugin, which installs index templates, component templates, and ingest pipelines for Elastic APM.

This replaces the need for installing any Fleet integration package for Elastic APM, unless central configuration/fleet management is desired, simplifying setup for users of the apm-server binary. This will also simplify APM Server by removing the possibility of index templates being missing on startup.

Commit:
x-pack/plugin/apm: introduce x-pack-apm plugin

felixbarny commented:
How do we ensure rollovers when settings or mappings change? Does this PR depend on https://github.com/elastic/elasticsearch/issues/96521?

axw commented:
@felixbarny we don't. We would either need #96521, or do it in the plugin:

https://github.com/elastic/elasticsearch/pull/97546/files#diff-652b74c665f6638e6113270681056a55bca105de6a22a297f83c0312634c841dR45-R47

> // TODO(axw) when the template version is updated, data streams using the old
> // version should be rolled over after upgrading. Ideally this would be done
> // automatically: https://github.com/elastic/elasticsearch/issues/96521

:smile:

ruflin commented on x-pack/plugin/apm/src/main/resources/index-templates/metrics-apm.app.yaml:
```diff
@@ -0,0 +1,19 @@
+index_patterns: ["metrics-apm.app.*-*"]
+priority: 150
+data_stream: {}
+composed_of:
```
@eyalkoren One more list for aligning with the convention

axw replied:
FYI, @felixbarny and I have discussed that in the (hopefully not too) long term, this index template should be removed and we should instead write metrics to a non-APM specific data stream, using the generic metrics template. That's why I added histogram & summary dynamic templates to `metrics-mappings.json`.

axw replied:
Same with `logs-apm.app`

eyalkoren replied:
@axw is working on renaming according to the [agreed naming scheme](https://github.com/elastic/elasticsearch/issues/96267)

axw commented:
@danielmitterdorfer provided some helpful pointers on Slack:

https://github.com/elastic/elasticsearch/pull/97380 adds support for upgrading Universal Profiling data streams and indices. Maybe we can take some inspiration from there to address https://github.com/elastic/elasticsearch/issues/96521?

https://github.com/elastic/elasticsearch/pull/97773 adds support for migrations. I haven't looked at all the details, but perhaps this could be useful for adding runtime fields to old backing indices for schema evolution.

eyalkoren commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IngestPipelineConfig.java:
```diff
@@ -50,15 +54,21 @@ public int getVersion() {
 
-    public BytesReference loadConfig() {
-        return new BytesArray(TemplateUtils.loadTemplate(resource, String.valueOf(version), versionProperty));
+    public PutPipelineRequest getPutPipelineRequest() {
+        return requestSupplier.get();
+    }
+
+    private static Supplier<PutPipelineRequest> newResourceSupplier(String id, String resource, int version, String versionProperty) {
+        Objects.requireNonNull(resource);
+        Objects.requireNonNull(versionProperty);
+        return () -> new PutPipelineRequest(
+            id,
+            new BytesArray(TemplateUtils.loadTemplate(resource, String.valueOf(version), versionProperty)),
+            XContentType.JSON
+        );
```
Nice convenience method, but I think out of scope here. This is a dumb class that only describes how the ingest pipeline resource is read from a file into memory, it doesn't know anything about the cluster or what a request is. I'd maintain this limited responsibility and keep it similar to its counterparts that serve the same purpose for other resources loaded by registries.

axw replied:
I've already paged out most of the details from my mind, but I think I added this so the pipeline resources didn't have to live in the x-pack-core plugin. Or maybe it was so the pipelines could be written as YAML. Or possibly both.

eyalkoren replied:
We can probably achieve both by providing an APM-specific subclass of `IngestPipelineConfig` with a custom `loadConfig()` overload. I'll see how it looks like, it may be better in terms of keeping this class's responsibility to only load the resource from disk.

eyalkoren committed:
Dependency fix and tests

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into apm-data-streams

eyalkoren committed:
Restore addition to ESRestTestCase

eyalkoren committed:
Replace IngestPipelineConfig instantiation

eyalkoren committed:
Update DataStreamUpgradeRestIT to expect logs-apm.error

eyalkoren committed:
Adding rollover functionality

axw commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java:
```diff
@@ -518,8 +529,14 @@ private void putComposableTemplate(
                     public void onResponse(AcknowledgedResponse response) {
-                        creationCheck.set(false);
-                        if (response.isAcknowledged() == false) {
+                        if (response.isAcknowledged()) {
+                            if (isUpgrade && applyRolloverAfterTemplateV2Upgrade()) {
+                                invokeRollover(state, templateName, indexTemplate, creationCheck);
```
Not sure if I was very clear when we chatted yesterday, so just reiterating in case: we need to consider that a failure may occur between the template update and rollover. If we don't cater for that, we'll end up with data streams in a half-updated state, and would require manual rollovers to resolve.

Could we make `invokeRollover` idempotent by comparing the active data stream backing index's applied template version to the current template version? Then we should be able to check if a rollover is needed on cluster state change, and be more resilient to failures.

If I'm reading the code properly, that's the approach taken in the profiling plugin: https://github.com/elastic/elasticsearch/blob/d1a2848bb3f69bb9990aa38fee99a544730efe75/x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/AbstractProfilingPersistenceManager.java#L75-L117

(Specifically see the `getIndexState` code.)

CC @danielmitterdorfer. We should probably align approaches so the profiling plugin could use this general logic in the future. WDYT?

eyalkoren replied:
> we need to consider that a failure may occur between the template update and rollover. If we don't cater for that, we'll end up with data streams in a half-updated state, and would require manual rollovers to resolve.

Note that we invoke a rollover only if the update is [acknowledged](https://github.com/elastic/elasticsearch/pull/97546/files#diff-a74931128493bfa2cd8c35c1bf3378014fd64c596c4bfbea0f91ddedd5db7011R532) by the cluster, which should guard from that IIANM.

> Could we make `invokeRollover` idempotent by comparing the active data stream backing index's applied template version to the current template version?

If I do that, I lose the existing synchronization mechanism that ensures that multiple updates of a template cannot happen concurrently. Follow the [`creationCheck`](https://github.com/elastic/elasticsearch/pull/97546/files#diff-a74931128493bfa2cd8c35c1bf3378014fd64c596c4bfbea0f91ddedd5db7011R516) and see how I utilized this mechanism so that index template updates and the corresponding rollovers happen atomically (in the context of the registry, not within the cluster).

danielmitterdorfer replied:
> CC @danielmitterdorfer. We should probably align approaches so the profiling plugin could use this general logic in the future. WDYT?

Thanks for the ping. In the profiling plugin we have chosen a different approach that decouples versioning of index templates from matching data streams (and indices) and their versioning. We do this because: 

* Not all changes to the index template require a rollover (e.g. removing a property in the template does not require a rollover).
* In some cases (e.g. for key-value lookup indices) rollovers are disruptive to cluster performance and disk usage.

Therefore we use `IndexTemplateRegistry` only to manage index templates (and ILM policies) and have introduced separate management classes (`ProfilingDataStreamManager` and `ProfilingIndexManager`) to manage the lifecycle of data streams / indices. These classes are inspired by `SystemIndexMappingUpdateService` that manages mapping updates for system indices. I wonder whether separating these two concerns (with some common infrastructure in the core plugin) would also make sense here?

felixbarny replied:
What we should really avoid is that every team solves the same problems in slightly different ways. As much as possible, we should try to contribute the enhancements from the APM and profiling plugins to the Elasticsearch core plugin.

felixbarny replied:
I think it's reasonable to defer an aligned solution in favor of iterating on different approaches up to a certain point. However, as these different solutions become more sophisticated, it will become increasingly complex to come up with a solution that satisfies all needs.

In this case, the profiling solution is already quite sophisticated (for example with migrations) and the apm solution doesn't seem to be adding lots of complexity. Also, none of the different solutions really expose something to users or have backwards compatibility guarantees, other than setting requirements on minimizing the number of rollovers. So I think I'm comfortable with moving along with something APM specific and deferring consolidation for a bit longer under the condition that the ES, APM, and the profiling teams commit to working on a common solution.

The first step of that would be to list the high-level requirements each team has to see if they're compatible and to evaluate the feasibility.

eyalkoren replied:
Just clarifying - what I added is indeed quite specific for the current basic APM requirements, but it is added as a general capability for any registry to opt-in to incur rollovers automatically upon index template upgrade. Probably not a functionality we will want to maintain once we have https://github.com/elastic/elasticsearch/issues/96521 or the like.

It is as simple as it gets, both on the positive and the negative side. The negative is of course that any registry update will trigger rollovers all over. Besides the obvious overhead related to that, it will probably also cause lots of excessive cluster state change events to be evaluated. Not sure this is an issue in itself, but worth being aware.

eyalkoren committed:
Extend basic rollover funtionality tests

eyalkoren committed:
Start adding integration test

Commit:
Hide rollup data streams

Commit:
Apm ingest fixes

- Map transaction.duration.us independently of span.duration.us,
  fix ingest pipeline. We might consider aliasing later.
- Set event.ingested
- Set processor.event, using constant_keyword where possible

Commit:
Fix error.grouping_name script

patpatpat123 commented:
cheering for this

Commit:
Only set event.ingested for traces-apm.sampled

eyalkoren committed:
Enabling APMRolloverIT

eyalkoren committed:
Merge remote-tracking branch 'axw/apm-data-streams' into apm-data-streams

eyalkoren committed:
Spotless...

eyalkoren committed:
Assertion change

eyalkoren committed:
Wait a bit before assertBusy

eyalkoren commented on x-pack/plugin/apm/src/internalClusterTest/java/org/elasticsearch/xpack/apm/APMRolloverIT.java:
```diff
@@ -0,0 +1,166 @@
+        // No need to waste the initial frequent assertBusy iterations because this takes a few seconds at least
+        Thread.sleep(3000);
+        assertBusy(() -> {
+            // todo - figure out why we need this synthetic cluster change event trigger to make sure the last index template is registered
+            // maybe because org.elasticsearch.ingest.IngestService.putPipeline() doesn't update the cluster state and trigger state
+            // change event like, for example, org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.addComponentTemplate does
```
My hypothesis is that in this case we don't install the [`metrics-apm.internal` index template](https://github.com/elastic/elasticsearch/blob/bac147f54cd5845d8d1593d9657f39917de610c9/x-pack/plugin/apm/src/main/resources/index-templates/metrics-apm.internal.yaml), because it depends on the [`metrics-apm.internal` component template](https://github.com/elastic/elasticsearch/blob/bac147f54cd5845d8d1593d9657f39917de610c9/x-pack/plugin/apm/src/main/resources/component-templates/metrics-apm.internal.yaml), which in turn depends on the [`metrics-apm.internal` ingest pipeline](https://github.com/elastic/elasticsearch/blob/bac147f54cd5845d8d1593d9657f39917de610c9/x-pack/plugin/apm/src/main/resources/ingest-pipelines/metrics-apm.internal.yaml) when the cluster-change-event gets processed.

@jbaiera do you think we missed this in https://github.com/elastic/elasticsearch/pull/95782? I don't think installation of ingest pipeline triggers a cluster change event, and for the registries to keep installing everything, it should.

jbaiera replied:
Ingest pipeline installation always triggers a cluster state update if the cluster is able to accept an ingest workload:

https://github.com/elastic/elasticsearch/blob/a20ee3f8f268a4f181b3e898635a96a3f605721a/server/src/main/java/org/elasticsearch/ingest/IngestService.java#L431-L435

eyalkoren replied:
Looking at the log of this test, the component template's is being added properly, but it may not be triggering the state change. See what happens if I remove the manual state change event (notice the timings):
```
[2023-09-13T04:18:15,091][INFO ][o.e.x.c.t.IndexTemplateRegistry] [node_t0] adding ingest pipeline metrics-apm.internal
[2023-09-13T04:18:15,093][INFO ][o.e.c.m.MetadataIndexTemplateService] [node_t0] adding component template [metrics-apm.internal]
[2023-09-13T04:18:27,224][INFO ][o.e.x.a.APMRolloverIT    ] [testRollover] [APMRolloverIT#testRollover]: cleaning up after test
```

And this is how it looks when I trigger such event within the `assertBusy` block, like the current code does:
```
[2023-09-13T01:22:55,447][INFO ][o.e.x.c.t.IndexTemplateRegistry] [node_t0] adding ingest pipeline metrics-apm.internal
[2023-09-13T01:22:55,449][INFO ][o.e.c.m.MetadataIndexTemplateService] [node_t0] adding component template [metrics-apm.internal]
[2023-09-13T01:22:57,714][INFO ][o.e.c.m.MetadataIndexTemplateService] [node_t0] adding index template [metrics-apm.internal] for index patterns [metrics-apm.internal-*]
```

There is definitely something wrong here, I'll debug later and see. If you have an idea- let me know

eyalkoren replied:
The issue went away after @axw refactored the components, so I removed the manual event trigger and resolve this thread, but I mean to restore the former state and figure out what the problem is. I think there is a real issue with cluster state events being missed in some configurations.

Commit:
Adjust template and pipeline names to convention

See https://github.com/elastic/elasticsearch/issues/96267#issuecomment-1662129512

- Index templates have a "@template" suffix
- Component templates are split into settings and mappings,
  and have a @settings and @mappings suffix respectively
- Ingest pipelines generally have a @pipeline suffix,
  except for one special case where we use @default-pipeline

We no longer have a component template for every data stream, rather we
have reusable component templates where it makes sense. For example, we
now have metrics-apm@settings and metrics-apm@mappings, which are
included by all APM metrics data streams.

We now set both default_pipeline and final_pipeline in the index
templates, preventing users from overriding them completely. The
default pipeline is always apm@default-pipeline, which performs
some rerouting of legacy data, and invokes user-defined @custom
pipelines based on the data_stream.type and data_stream.dataset
fields. The final pipeline just performs built-in processing.

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into apm-data-streams

eyalkoren committed:
No need for manual cluster change events anymore

eyalkoren committed:
Fix test

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into apm-data-streams

Commit:
Lower template priority to 110

Certain index templates installed by Fleet are given
a priority of 150, so lower the priority to avoid
conflicts.

See https://github.com/elastic/kibana/blob/a18c68f7ac0b2cb6a46552328e9c6fc28c223970/x-pack/plugins/fleet/server/services/epm/elasticsearch/template/template.ts#L58

eyalkoren committed:
Add logging

Commit:
Add _meta to component templates too

eyalkoren committed:
Remove explicit timeout for rollovers

eyalkoren committed:
Merge remote-tracking branch 'axw/apm-data-streams' into apm-data-streams

Commit:
Increase template priorities to 140

This goes closer to the Fleet-installed template priorities,
allowing for more builtin templates to be introduced without
affecting APM.

jbaiera reviewed (commented):
I think there might be a problem here if we only check that the composable template has been updated.

Looking at the code, it may be possible that the composable templates for a data stream have been updated but the component templates it depends on have not yet. This would be unlikely to cause an issue during normal operation due to natural timing, but if a component template fails to update correctly due to a merging issue, then the data stream could rollover before the component template is correctly updated, leading to a manual intervention being required.

eyalkoren commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java:
```diff
@@ -405,7 +415,7 @@ private void addComposableTemplatesIfMissing(ClusterState state) {
                         newTemplate.getValue().version()
                     );
-                    putComposableTemplate(templateName, newTemplate.getValue(), creationCheck);
+                    putComposableTemplate(state, templateName, newTemplate.getValue(), creationCheck, true);
```
That's a very good point, but maybe since we are aiming to rely on https://github.com/elastic/elasticsearch/issues/96521 instead of this rollover mechanism in the long term, we can either decide to manually keep track of that, or simply decide we increase the version in the APM registry in every release until we have the full solution.
I think the bigger problem will be with the combination of the issue you raised below, where we need to know about dependencies versions as well...

@axw how is it with Fleet-managed index templates? Would Fleet know to rollover a data stream if its corresponding index template did not change, but it uses a stack-managed templates (like `logs-mappings`) that gets changed?

axw replied:
@eyalkoren Fleet does not add any stack-managed templates to the index templates it installs at the moment. So it's not a problem there.

IMO we should defer solving this to #96521

felixbarny replied:
It doesn't do that at the moment but it soon will: https://github.com/elastic/kibana/issues/163141.

> Will there be a mechanism for rolling over a data stream if a component template changes?

IIRC a discussion with @axw and @eyalkoren we said that for now, we're fine with having to manually bump the version of the index template if a referenced component template changes.

https://github.com/elastic/elasticsearch/issues/96521 is the long-term solution.

eyalkoren replied:
I edited my comment to explain this better. The APM registry uses the same version for all assets.
As @jbaiera implied, the problem is with templates that are coming from other registries

eyalkoren commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java:
```diff
@@ -518,8 +530,14 @@ private void putComposableTemplate(
                     public void onResponse(AcknowledgedResponse response) {
-                        creationCheck.set(false);
-                        if (response.isAcknowledged() == false) {
+                        if (response.isAcknowledged()) {
```
That's an excellent point!! We'll need to enhance the dependencies validation mechanism to be capable to look at component templates versions as well. I think it should be doable if both the dependencies and the dependents are coming from the same registry, but not so sure otherwise...

I'll try to think of something to work around this issue

eyalkoren replied:
Part 1: I extended the dependencies check that is done during index template installation to look at component templates' versions as well (see https://github.com/elastic/elasticsearch/pull/97546/commits/665dc9de880b247a590b853a3f6600579c373f60). 
Currently this check includes only component templates that are installed by the same registry that in installs the index template
This would not affect any registry that doesn't require automatic rollovers (meaning - has no effect on any other than the APM registry).

Part 2: I am now looking for a way to apply this also to component templates that come from other registries. It sorts of introduces a concept of inter-registries dependency, which I am not sure is desirable. I will try to propose something if I find a proper way. 
@jbaiera if any of you have a concrete idea, please let me know.

eyalkoren replied:
@jbaiera since this plugin is going to be disabled by default for now and since the worst outcome is of premature rollover, we decided to handle "Part 2" above as a followup for this PR (see #99647). 
Please do a comprehensive review for this PR assuming that this external dependencies upgrade validation is out of scope.

eyalkoren committed:
Test multiple index template upgrades with rollovers

eyalkoren committed:
Enfore correct component template versions

eyalkoren committed:
Merge remote-tracking branch 'axw/apm-data-streams' into apm-data-streams

rjernst commented:
We already have an apm module in Elasticsearch (for pushing data to otel/apm server). Could we name this xpack module differently, to distinguish the two? Perhaps `apm-server`?

Commit:
Rename to x-pack-apm-data, default to disabled

Because there's already an "apm" plugin,
we haven't yet solved all issues with rollovers,
and the templates need some more love.

axw commented:
> @rjernst we could rename it to apm-data for some symmetry with https://github.com/elastic/apm-data, which is what defines our data model and ES document encoding. Does that sound OK to you?

I went ahead and renamed it to `x-pack-apm-data`, renamed the enablement setting to `xpack.apm_data.enabled`, and switched it to disabled by default.

Commit:
Fix checkstyle

Commit:
Fix more checkstyle

Commit:
Fix even more linting issues

eyalkoren committed:
Adjust test to disabled by default

eyalkoren committed:
remove todo

eyalkoren committed:
Fixing test

eyalkoren committed:
Refactor leftovers

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into apm-data-streams

rjernst commented:
> I went ahead and renamed it to x-pack-apm-data

I don't have a better name for now.

eyalkoren committed:
Fix constant keyword mapping conflict

jbaiera commented on x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycle.java:
```diff
@@ -179,6 +179,7 @@ public Collection<Object> createComponents(
             xContentRegistry
         );
         ilmTemplateRegistry.initialize();
+        components.add(ilmTemplateRegistry);
```
Bookkeeping note: As mentioned in my comment in the APMPlugin class, if we move the APMRolloverIT to core, this doesn't need to be exposed.

jbaiera commented on x-pack/plugin/apm-data/src/internalClusterTest/java/org/elasticsearch/xpack/apmdata/APMRolloverIT.java:
```diff
@@ -0,0 +1,169 @@
+
+@SuppressWarnings("resource")
+// the Painless compiler requires the "createClassLoader" permission
+@ESTestCase.WithoutSecurityManager
```
This strikes me as something that needs to be cleaned up before the PR can be accepted - There are only a small handful of tests that need this currently and they seem to be much lower level than what this is testing. Is there a way we could solve this problem some other way? I'm not entirely convinced that this test case makes sense where it is, but if we keep it around, I think this would need a more proper fix.

eyalkoren replied:
I agree. I looked for something, assuming there is a test setup I need to make to prevent this, but couldn't find one so far. I'll keep looking.

jbaiera reviewed (commented):
I've added a number of suggestions for the sake of code quality and clarity. There are a few changes here that could be streamlined a little bit. Happy to discuss them further if they raise any concerns.

There is one larger issue that I noticed upon looking at this again: We manage different permissions for rolling over a data stream and for changing index templates in a cluster. With these changes, it is possible for a user with access to index templates to trick the registry into performing a rollover by putting an old version of the index template in place, even if said user does not have permissions to rollover the data streams affected. This seemed to me to be unexpected behavior. I imagine the overall impact of this is low, but it could lead to unintended consequences in the wild.

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into HEAD

eyalkoren committed:
Fix after merge

eyalkoren commented on x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/ResourceUtils.java:
```diff
@@ -0,0 +1,38 @@
+
+    public static final String APM_TEMPLATE_VERSION_VARIABLE = "xpack.apmdata.template.version";
+
+    static byte[] loadVersionedResourceUTF8(String name, int version) {
```
I [extracted it](https://github.com/elastic/elasticsearch/commit/cc750adf843a823c5fe86517239577ffa4fd8450) from the `ApmIngestPipelineConfig`, I'll see if I can use `TemplateUtils` instead.

> TemplateUtils has a validation step. It uses XContent but that could also be parameterized.

Maybe this validation is what made it less fit in the first place, as it assumes JSON files and the APM-data plugin uses yaml files

eyalkoren replied:
Looking back on this, I think it makes no sense to change `TemplateUtils` to accommodate that. `ResourceUtils` is extremely lean. The only think I thinks worth reusing is `TemplateUtils.replaceVariable()`, which it already does. Other than that, it's code is minimal, not nothing worth refactoring in order to share IMO.
- the validation in `TemplateUtils` is irrelevant as it specific to JSON format
- the resource loading is really very minimal and we cannot use `TemplateResources#load`, unless we move all resources to the core resources directory, which we don't want

eyalkoren commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IngestPipelineConfig.java:
```diff
@@ -17,11 +17,11 @@
-    private final String resource;
-    private final int version;
-    private final String versionProperty;
+public abstract class IngestPipelineConfig {
```
I never thought of subclassing as heavyweight, but as a very basic language feature.

> The only differences between the other subclasses is which XContent type it returns (could be a member variable passed in via the constructor instead) and how the data is loaded (could be specified with a supplier function when the config is created instead of needing a subclass).

To me this sounds like you prefer to implement polymorphism while avoiding the language's most basic concept for polymorphism (inheritance). I can't see how what you propose is better that inheritance, especially in terms of readability, with class names indicate exactly what they differ at and method overloads. However, if there is tendency in this code base to avoid class hierarchies when possible, I will follow suit.

eyalkoren replied:
Maybe if I rename the `ApmIngestPipelineConfig` to `YamlIngestPipelineConfig` it would make more sense. I prefixed it with "Apm" as it is located within the APM plugin and currently doesn't serve as a general-purpose subclass, but naming it consistently with the other subclass (`JsonIngestPipelineConfig`) will probably clarify more what this class hierarchy is about.

jbaiera replied:
In my experience it is better to value composition over inheritance, especially in cases such as these where the subclasses are primarily providing values to the main class instead of providing substantial changes in behavior.

For instance, the subclasses each have a method that returns a constant value of which XContent they are meant to handle. That constant could just be passed in as part of the object's constructor. The most significant difference is how they load data, and even that is just a small difference in which format the data is provided in. That could easily be abstracted out of the class with a function or supplier instead of creating an object hierarchy on the ingest pipeline config.

Ultimately it's probably not worth a long discussion. My feedback is admittedly a nitpick, and my concern is primarily structural instead of how the code is behaving. I tend to have a really suspicious response to changes that introduce an object hierarchy.

eyalkoren replied:
I see what you're saying and I generally agree.
I [reverted](https://github.com/elastic/elasticsearch/pull/97546#discussion_r1297064084) @axw's change of this class to make sure it stays with a single functionality, which is knowing how to load the underlying file. Maybe this is why we see it differently - to me this is ALL these classes are doing. 
Everything in core uses JSON pipeline files that are served from the same location (which I think is a limitation by itself), so I think it's easier for all users to pick a type instead of specifying the same argument in constructor and a supplier.

eyalkoren commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java:
```diff
@@ -501,9 +536,11 @@ public void onFailure(Exception e) {
         final ComposableIndexTemplate indexTemplate,
-        final AtomicBoolean creationCheck
+        final AtomicBoolean creationCheck,
+        final boolean isUpgrade
```
This is not a change related to this PR, it is how this registry is handling concurrency across all component upgrades.

> do we need to keep the flag true for the entire execution of the rollover

I believe we do, otherwise we may get into concurrent rollover request for the same data stream. I don't know what the exact consequences of that, but definitely sounds like something we better avoid. What's the down side? If a composable template gets upgraded and we find data streams that are affected by it, why should we allow for additional upgrade **for the same composable template** before rollovers are done?

Eventually, our hope is to have this as a basic Elasticsearch capability (like proposed in  https://github.com/elastic/elasticsearch/issues/96521#issuecomment-1695801781) that the registries only opt-in for. I assume that any such implementation will use the same atomicity semantics, where the composable template upgrade action includes the subsequent rollovers as part of it before it can be considered done.

jbaiera replied:
> I believe we do, otherwise we may get into concurrent rollover requests for the same data stream.
How might we end up in that scenario? This would require the composable template to be changed and for the template installation process to kick off again correct?

Again, this comment is mostly seeing if we can improve the structure of the code. If that comes at a cost to the correctness of the behavior then we can ignore the suggestion.

eyalkoren replied:
This is mainly where I need your input from the review - is the code **correct**? 
My assumption is that we keep the former behavior of avoiding concurrent index template upgrades, and that rollovers of affected data streams should be considered now part of the index template upgrade. Is this assumption correct? Is the code properly enforcing that?

jbaiera replied:
This code does seem to be correctly enforcing that. My comments here were just questioning if waiting for rollovers to complete before marking the template install complete was something that made sense to do or not. For now, I think it's fine to have that be the case.

eyalkoren commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IndexTemplateRegistry.java:
```diff
@@ -732,4 +790,91 @@ public void onFailure(Exception e) {
+        executor.execute(() -> {
+            List<String> rolloverTargets = findRolloverTargetDataStreams(state, templateName, indexTemplate);
+            if (rolloverTargets.isEmpty() == false) {
+                GroupedActionListener<RolloverResponse> groupedActionListener = new GroupedActionListener<>(
```
See my [reply](https://github.com/elastic/elasticsearch/pull/97546#discussion_r1361483834) above.
If we decide that there's no good reason to block composable templates upgrade before all affected data streams are rolled over, we can change this

eyalkoren commented on x-pack/plugin/apm-data/src/main/java/org/elasticsearch/xpack/apmdata/APMPlugin.java:
```diff
@@ -0,0 +1,80 @@
+        registry.set(new APMIndexTemplateRegistry(settings, clusterService, threadPool, client, xContentRegistry));
+        if (enabled) {
+            registry.get().initialize();
+            return List.of(registry.get());
```
Thanks for the pointer, I will update to behave like the `StackTemplateRegistry`, which is always returned and looks at the `enabled` state internally.

eyalkoren commented on x-pack/plugin/apm-data/src/internalClusterTest/java/org/elasticsearch/xpack/apmdata/APMRolloverIT.java:
```diff
@@ -0,0 +1,169 @@
+        );
+    }
+
+    public void testRollover() throws Exception {
```
You're absolutely right! I located this test in the APM plugin for two reasons:
1. this plugin already contains many different examples of inter-component dependencies (within its registry assets). Testing that the rollover occurred also implicitly verifies that dependencies verification do not interfere with the composable template upgrade. Testing it here was a huge shortcut in trying to mock all these dependencies and whether they affect the rollover.
2. nothing else at the moment fully tests APM data stream creation, upgrades and eventually rollover. I completely agree that this is much more suitable for REST integration tests (i.e. ingest a sample document, test that a data stream is created, test mappings etc.). I discussed this briefly with @axw - we should have a much more extensive yaml REST integration test suite, but at the moment what [we have](https://github.com/elastic/elasticsearch/pull/97546/files#diff-efcce7628d0dd23f916e701bc4632e923410bfc3c949ba290aefe332f9ab4f2f) is quite minimal.

I made my changes to this PR with MVP concerns in mind mostly, including this. I think we will have subsequent PRs to complete the full scope, part of which is adding a proper test suite, where we can also address this once we know we have integration tests that create all data steams. WDYT?

eyalkoren replied:
Removed this integration test and added one to core instead

eyalkoren committed:
Comment fixes

eyalkoren commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackSettings.java:
```diff
@@ -92,6 +92,13 @@ public Iterator<Setting<?>> settings() {
     );
 
+    /** Setting for enabling or disabling APM Data. Defaults to false. */
+    public static final Setting<Boolean> APM_DATA_ENABLED = Setting.boolSetting(
```
Note that part of the [plan](https://github.com/elastic/elasticsearch/issues/98535#issuecomment-1687818059) of the component template refactor is to address that, specifically the task that reads: 

" add a new stack template registry that applies the `stack.templates.enabled` flag (defaults to `true`) only for index templates- if set to `false`, only index templates are not registered, component templates are still registered"

This should remove this dependency.

eyalkoren committed:
Rename ApmIngestPipelineConfig to YamlIngestPipelineConfig

eyalkoren committed:
Always return a registry

eyalkoren committed:
spotless, what else

eyalkoren committed:
Add rollover integration test to core

eyalkoren committed:
Remove APMRolloverIT

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into apm-data-streams

eyalkoren commented:
@jbaiera I addressed all review comments. As agreed with @axw and @felixbarny, full integration tests for APM components will be added in a followup effort. I removed the rollover test from the APM data plugin and added a simple one that tests the basic functionality to core instead.
Please go over the non-resolved comments and see if you can resolve them. Thanks for your great feedback 🙏

eyalkoren committed:
Merge remote-tracking branch 'upstream/main' into apm-data-streams

eyalkoren committed:
Refactor in renamed components

jbaiera reviewed (approved):
Thanks for iterating on this! Given that the index template registry refactoring is on the todo list for future consideration, I think things look good to me!

axw commented:
Thank you for finishing this up @eyalkoren, and for your support @jbaiera! :heart:

patpatpat123 commented:
Hello team,

Thank you for the work on this.

I am using the latest elastic, kibana, apm server, at version 8.11.3.

With that set up, I am still seeing :

```
 Error loading integration details

Kibana security must be enabled to use Fleet
```
```
Unable to check permissions

There was a problem checking Fleet permissions
  ```
  
  ```
  Error loading integration details

Kibana security must be enabled to use Fleet
```
```
[2023-12-30T16:55:47.984+08:00][INFO ][plugins.fleet.fleet_authz_router] Kibana security must be enabled to use Fleet
[2023-12-30T16:55:48.001+08:00][INFO ][plugins.fleet.fleet_authz_router] Kibana security must be enabled to use Fleet
[2023-12-30T16:55:48.017+08:00][INFO ][plugins.fleet.fleet_authz_router] Kibana security must be enabled to use Fleet
[2023-12-30T16:55:48.026+08:00][INFO ][plugins.fleet.fleet_authz_router] Kibana security must be enabled to use Fleet
  
```

How to use xpack apm plugin, instead of this fleet based apm?

Thank you

tvernum commented:
Hi @patpatpat123, thanks very much for your interest in Elasticsearch.

This appears to be a user question, and we'd like to direct these kinds of things to the [Elasticsearch forum](https://discuss.elastic.co/c/elastic-stack/elasticsearch/6). If you can stop by there, we'd appreciate it. This allows us to use GitHub for verified bug reports, feature requests, and pull requests. The discussion forums are better suited to these sorts of product usage conversations.

There's an active community in the forums that should be able to help get an answer to your question. As such, I hope you don't mind that I hide your comment from this ticket.

Thank you.

edzhelyov commented:
Sorry to jump on an already closed ticket, but I'm struggling to get a local Docker APM server running without me understanding the whole Kibana ecosystem.

The latest (8.12) documentation still suggest using `xpack.fleet.packages`, but this doesn't work with insecure configuration (whatever that means).

Is this x-pack/plugin available in 8.12 to use, or do I have to enable the security?

DaveCTurner commented:
Hi @edzhelyov, thanks very much for your interest in Elasticsearch.

This appears to be a user question, and we'd like to direct these kinds of things to the [Elasticsearch forum](https://discuss.elastic.co/c/elasticsearch). If you can stop by there, we'd appreciate it. This allows us to use GitHub for verified bug reports, feature requests, and pull requests.

There's an active community in the [forum](https://discuss.elastic.co/c/elasticsearch) that should be able to help get an answer to your question. As such, I hope you don't mind that I am hiding your comment on this ticket as off-topic.
